File: cs/Problems/Day7.cs
namespace AOC2023.Problems;

public class Day7 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => CalculateTotalCardWinnings(input);

    private static int CalculateTotalCardWinnings(ReadOnlySpan<char> input)
    {
        Span<Range> lineRanges = stackalloc Range[1000];

        int lines = input.Split(lineRanges, InputReader.NewLine);
        Span<CardDraw> draws = stackalloc CardDraw[lines];

        for (int i = 0; i < lines; i++)
            draws[i] = CardDraw.Parse(input[lineRanges[i]]);

        draws.Sort(CardDraw.Compare);

        int totalWinnings = 0;

        for (int rank = 1; rank <= lines; rank++)
            totalWinnings += draws[rank - 1].Bid * rank;

        return totalWinnings;
    }

    readonly record struct CardDraw(long DrawValue, int Bid)
    {
        const string CARD_TYPES = "23456789TJQKA";

        public static CardDraw Parse(ReadOnlySpan<char> cardDrawString)
        {
            int separator = cardDrawString.IndexOf(' ');

            var draw = cardDrawString[..separator];
            int bid = int.Parse(cardDrawString[(separator + 1)..]);
            long drawValue = CalculateDrawValue(draw);

            return new CardDraw(drawValue, bid);
        }

        public static int Compare(CardDraw first, CardDraw second) => first.DrawValue.CompareTo(second.DrawValue);

        static long CalculateDrawValue(ReadOnlySpan<char> drawnCards)
        {
            int typeCount = 0;
            int maxCount = 1;
            int currentCount = 1;
            long drawValue = 0;

            var cards = drawnCards.ToArray();
            Array.Sort(cards);

            for (int i = 0; i < cards.Length; i++)
            {
                if (i > 0 && cards[i] == cards[i - 1])
                {
                    currentCount++;
                }
                else
                {
                    typeCount++;
                    currentCount = 1;
                }

                maxCount = Math.Max(maxCount, currentCount);
                drawValue = (drawValue * 100) + (CARD_TYPES.IndexOf(drawnCards[i]) + 10);
            }

            int drawType = typeCount switch
            {
                1 => 7,
                2 when maxCount is 4 => 6,
                2 => 5,
                3 when maxCount is 3 => 4,
                3 => 3,
                4 => 2,
                _ => 1
            };

            return (drawType * 10_000_000_000) + drawValue;
        }
    }
}

File: ts/problems/day7.ts
import { NEW_LINE } from "../inputReader.js";

const calculateTotalCardWinnings = (input: string) =>
  input
    .split(NEW_LINE)
    .map(parseCardDraw)
    .sort(sortByDrawValue)
    .reduce((sum, _, i, draws) => sum + draws[i].bid * (i + 1), 0);

const parseCardDraw = (cardDrawString: string): CardDraw => {
  const [draw, bidValue] = cardDrawString.split(" ");
  const drawValue = calculateDrawValue(draw);
  return { draw, drawValue, bid: Number(bidValue) };
};

const calculateDrawValue = (drawnCards: string) => {
  let maxCount = 1;
  let currentCount = 1;
  let typeCount = 0;
  let drawValue = 0;

  [...drawnCards].sort().forEach((_, i, cards) => {
    currentCount = i > 0 && cards[i] === cards[i - 1] ? currentCount + 1 : 1;
    typeCount = currentCount === 1 ? typeCount + 1 : typeCount;

    maxCount = Math.max(maxCount, currentCount);
    drawValue = drawValue * 100 + (CARD_TYPES.indexOf(drawnCards[i]) + 10);
  });

  // prettier-ignore
  const drawType =
    (typeCount === 1) ? 7 :
    (typeCount === 2 && maxCount === 4) ? 6 :
    (typeCount === 2) ? 5 :
    (typeCount === 3 && maxCount === 3) ? 4 :
    (typeCount === 3) ? 3 :
    (typeCount === 4) ? 2 : 1;

  return drawType * 10_000_000_000 + drawValue;
};

const sortByDrawValue = (a: CardDraw, b: CardDraw) => a.drawValue - b.drawValue;

const CARD_TYPES = "23456789TJQKA";

type CardDraw = {
  draw: string;
  drawValue: number;
  bid: number;
};

export default calculateTotalCardWinnings;
